### PR TITLE
Batch support

### PR DESCRIPTION
**Pull request by crufter**
Highlights:
- All DML (Set, Update, Delete) and Read methods now return an `Op` type (recipe tables and the CQL one included).
- Instances of `Op` can be run together, either with `Run` which is equivalent to 'Run these however you want I don't care' and `RunAtomically` which means 'Run the DML statements in an atomic way ie. either all of them happens or none of them`.
- This is a breaking change.

Known limitations:
- `QueryExecutor` RunAtomically() does not actually run things atomically. This will be fixed once the interface changes are accepted.

### CODE REVIEW

**Comment by crufter** [comment]
Seems like I will introduce the concept of WriteOps to make batching first class. @andreas reviewed the idea and approved. @andreas suggested using a fluid, chainable API for batching.

Will explore this a bit later.

**Comment by andreas** [comment]
What do you think about having `Together` and `Atomically` on `WriteOp`? To me it feels a little weird with `Batcher` being embedded in `Connection`, `KeySpace` and all recipes.

``` go
type WriteOp interface {
  Add(...WriteOp) WriteOp
  Run() error // Together
  RunBatched() error // Atomically
}

users.Where(Eq("Tag", "hailo")).Update(map[string]interface{}{
  "Discount": 12.5,
}).Add(users.Where(Eq("Tag", "friend-of-hailo")).Update(map[string]interface{}{
  "Discount": 5.0,
}).RunBatched()
```

**Comment by crufter** [comment]
I really like those ideas, pretty much all of them! Having Batcher being embedded everywhere rubs me in the wrong way as well.

I like how Run invokes a feeling of being cheaper than RunBatched, unlike Together and Atomically.

@jcampan talked me into not promoting the WriteOps as the way to do it because people will overuse logged batches (which are really expensive). However, I think I will go with them after all and I will nuke the whole Reader Writer interface distinction. With the Together/Atomicatlly or Run/RunBatched we can give people the choice if they want to heavier or lighter 'batch', unlogged batch being the default, advised one.

Now I am only bothered about the asymmetry of the WriteOps versus the nonexistent ReadOps.
To be honest now that we unmarshal into input arguments a possible ReadOps implementation became easier.

Cassandra only supports batched writes (http://www.datastax.com/documentation/cql/3.1/cql/cql_reference/batch_r.html) at the moment but it doesn't mean we can not express it in our API. I suspect though since reads are much slower anyway the query time / network roundtrip spared ratio will be much higher for read queries than for writes.

**Comment by andreas** [comment]
Can you give a quick example of what ReadOps would look like?

**Comment by crufter** [comment]
Yes:

``` go
c1 := &Customer{}
err := customers.Read("1", c1)
if err != nil {
   return err
}
c2 := &Customer
err = customers.Read("2", c2)
if err != nil {
   return err
}
```

vs

``` go
c1 := &Customer{}
c2 := &Customer{}
err := customers.Read("1", c1).Add(customers.Read("2", c2)).Run()
if err != nil {
   return err
}
```

Just realized while typing that it has the benefit that it does not force you to check the errors after each query, haha. However, with longer argument list this nesting will not be as quite neat anyway, it will be more like:

``` go
c1 := &Customer{}
rOp1 := customers.Read("1", some, really, long, argument, list)
c2 := &Customer{}
rOp2 := customers.Read("2", something potentially even longer)
err := rOp1.Add(rOp2).Run()
if err != nil {
   return err
}
```

I admit this is a bit more controversial than batching writes, but I thought I mention it anyway.

**Comment by andreas** [comment]
One the one hand I like the symmetry of `WriteOp` and `ReadOp`, as well as only handling one error instead of _n_; on the other hand it seems a little peculiar or uncommon to "run" your reads.

**Comment by crufter** [comment]
+1

There is one other point I like: it nicely captures that reads happen at the same time, although this information is not useful at the moment.

From a user point of view probably this would be more confusing than useful so let's ignore ReadOps for the moment and focus on WriteOps.

So for the following reasons I will go with WriteOps in this PR, even in the face of @jcampan advise against doing so:
- Two different methods for running WriteOps (Run vs RunBatch) will mean we do not advocate/force users to use logged batches
- WriteOps come with less complexity in the API than writing different interfaces for Batched Write interfaces (which work well with 'flat' recipe table interfaces but don't mesh nicely with more layered interfaces like the Raw Table table set/filter update delete)
- Calling Run() on a WriteOp is a small enough price to pay (this is subjective though)

File: writeop.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	}
+	return w.qe.ExecuteBatched(stmts, params)
+}
+
+func (w writeOps) Add(wo ...WriteOp) WriteOps {
+	return writeOps {
+		qe: w.qe,
+		ws: append(w.ws, wo...),
+	}
+}
```
Comment (andreas) [inline writeop.go]: Does it make sense to type-wise distinguish between WriteOp and WriteOps?

File: writeop.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	ws []WriteOp
+}
+
+func (w writeOps) Run() error {
+	for _, v := range w.ws {
+		err := v.Run()
+		if err != nil {
+			return err
+		}
+	}
```
Comment (andreas) [inline writeop.go]: Can all of them be executed in one go (not batched but with one roundtrip)?

**Comment by crufter** [comment]
So at the moment my main concern is:

``` go
// Set will return a WriteOp and not an error actually
err := tbl.Set(Customer{
   Id: "2",
   Name: "I am not called Joe anymore",
})
// WriteOp won't be a nil
if err != nil {
   // Freak out without using err, thus not realizing err is in fact a WriteOp?
}
```

The above scenario won't be catched by the compiler. It might be a contrived one however.

File: writeop.go (hunk 2 above)
Comment (crufter) [inline writeop.go]: I think there is no way/makes no sense to do that, but I am not sure pinging @jcampan here.

Comment (andreas) [inline writeop.go]: Could it be a matter of joining the individual CQL statements with `;`?

File: writeop.go (hunk 1 above)
Comment (crufter) [inline writeop.go]: Well the only reason I can think of is calling RunBatched on a single WriteOp makes no sense, however we might just recognise it's called on one WriteOp and use Run instead?

File: writeop.go (hunk 2 above)
Comment (crufter) [inline writeop.go]: Good idea, will ask @jcampan. According to @jcampan we won't win much by doing so but I am still keen on reducing the network roundtrip n -> 1

**Comment by jcampan** [comment]
Just a quick remember about our last talk. 
I think that is should be better to do individual call for each query instead to joining them into one query. 
In this way, we don't need to take care of unbalanced traffic and all these details that cassandra handle already. An important point should be to have a good connection pool manager and avoid to open new connection for each query.
We need to see this performance before to try any optimisation (or we will never see if what we will try after is better or worst than the simple solution)

**Comment by crufter** [comment]
Thanks, @jcampan.
What do you think about the current implementation of seemingly running multiple WriteOps at the same time but the implementation actually just executes the DML queries one by one?

**Comment by obeattie** [comment]
@jcampan Please see connection pooling I added in hailocab/go-service-layer#220. I'd appreciate your review.

**Comment by jcampan** [comment]
The .run() and .runBatch() concept sounds good to me.   My only concern is about rename  the runBatch() in something like runSlowAtomicBatch() :P . I would like to make it clear that have an atomic batch is killing the performance and we should use it only if we really need atomic multi write operation.

**Comment by crufter** [comment]
@jcampan We can make that _very_ clear in the documentation, I don't think we have to resort to including 'slow' in the function name :D.

**Comment by crufter** [comment]
@braceta As one of the first users of the library, what do you think about this PR?

**Comment by braceta** [comment]
@crufter regarding this PR I like the chaining of methods for batched operations. That way you just need to handle error in one place. Looks great!

What about the Read op? Should it return a ReadOp in the same way so that you have symmetry?

Btw, one of the thing I ran into was when the Read doesn't found the record for a given key it returns an err. My view on it is that it should probably have the same behaviour as when you fetch an element from a map: 

```
someStruct = &SomeStruct{}
err, found :=  someOnetoOnetable.Read(key, someStruct)
```

or just set `someStruct` to nil (don't even know if Go allows it).

Because on a real error you don't want to try to recover a missing value.

**Comment by crufter** [comment]
@braceta Yes, in this thread there is a lot of talk about wether having `ReadOp`s is a good idea. Have you read those comments? What is your preference?

You are very right about Read lacking a way to differentiate between 'not found' and 'read error'

Possible solutions for this:
- Setting passed in pointer to nil (as you proposed). Yes it is possible (see http://play.golang.org/p/SFW0_noEqu) but I am afraid we would run into null pointer issues.
- Having an Exists method which checks for existence - for me this seems a bit superfluous.
- Returning an bool (as you proposed), this is nice and explicit, but if someone considers lack of Row and error they have to check for `!found || err != nil`.
- Returning a `RowNotFound` error type people can check for, this may not be obvious at first though.

I prefer the last one.
Let me know if you have know other ways.

**Comment by jcampan** [comment]
@crufter " We can make that very clear in the documentation,"
That is not an answer, we already talk a lot about it. The situation where we were with gossie and cassandra is because people are not reading documentation. In a nice API you dont need to read documentation  to understand that you are doing something witch is going to kill performance...
So runBatch is not clear enough.

A batch is not a nice way to perform multiple write, it s a solution to have an atomic write sequence and the cost is very expensive for that.

**Comment by obeattie** [comment]
I like this :+1:

**Comment by crufter** [comment]
@jcampan All right, so the original name for batch was 'Atomically' for this reason, do you like that better? Is RunAtomically enough? I am not sure if conveying slowness in a function name is appropriate or possible at all without resorting to inelegant solution like 'RunSlowBatchDontUseThisUnlessYouKnowWhatYouAreDoing', hence my suggestion of including this in the documentation.

So what are your suggestions for naming this method?

Also:

> In a nice API you dont need to read documentation to understand that you are doing something witch is going to kill performance...

Can you show an example of this?

**Comment by crufter** [comment]
Note: change RunBatched to RunAtomically.

**Comment by jcampan** [comment]
Thanks for this change.

File: gocql_backend.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+func (cb goCQLBackend) Execute(stmt string, vals ...interface{}) error {
+	return cb.session.Query(stmt, vals...).Exec()
+}
+
+func (cb goCQLBackend) ExecuteAtomically(stmts []string, vals [][]interface{}) error {
+	if len(stmts) != len(vals) {
+		return errors.New("executeBatched: stmts length != param length")
+	}
+	// THIS IS A FAKE TEMPORAL IMPLEMENTATION
```
Comment (andreas) [inline gocql_backend.go:32]: Nitpick: temporal => temporary

**Comment by andreas** [comment]
Thoughts:
- I really like having a unified `Op` rather than `ReadOp` and `WriteOp`. The name is a little opaque though.
- I would consider not distinguishing between `Op` and `Ops` (i.e. the type of `Add` could be `func (...Op) Op`). I think we've discussed this before. :)
- The failure semantics are not 100% clear from the API. Does `op1.Add(op2).Run()` execute `op2` if running `op1` returns an error (i.e. do you stop or continue on error)? Maybe the returned error should provide more information to the caller (which ops succeeded, which failed, which remains).

**Comment by crufter** [comment]
Good points.

What is the drawback of having an `Op` and `Ops`? The upside is I think it doesn't let you call Atomically on a single `Op` (more type safe).
Failure semantics - I thought about creating an other PR sorting out the errors everywhere. So yes, now it is a bit fuzzy.

**Comment by andreas** [comment]
In my mind, the simplicity of only having `Op` is more important than what I consider to be a rare edge case (batch exec'ing a single statement). `Op` could even be clever enough to not batch exec if it only contains 1 statement.

``` go
func (w *op) RunAtomically() error {
  if len(w.stmts) == 1 {
    return w.Run()
  }
  // ...
}
```

**Comment by crufter** [comment]
Since I don't have a strong counterargument, only a hunch, I will go with the `Op` then - since it is simpler, it is the more complex solution which needs defending not the simpler one.

**Pull request merged**